### PR TITLE
Add 1D versions of AMR positivity limiter

### PR DESCRIPTION
**Pull request by andrewwinters5000**
As discussed in #164 in [this thread](https://github.com/trixi-framework/TrixiShallowWater.jl/pull/164#discussion_r3557910110) the AMR capability is available for `TreeMesh1D` because no special mortar treatment is needed. It only requires new positivity limiter functions that ensure a positive water height after coarsening or refinement.

The first result result below is a beach run-up with four refinement levels found in `examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl` at time `t = 1.25` is given below. The Löhner indicator is quite sensitive in the dry regions and a specialized AMR indicator may need implemented (outside the scope of this PR).

The second result below is a comparison with the fixed 128 element mesh version from `examples/tree_1d_dgsem/elixir_shallowwater_beach.jl` (dashed green line) against the AMR version (solid blue line) with element breakdown
```
 #elements:                  75
 ├── level 8:                40
 ├── level 7:                 6
 ├── level 6:                 7
 └── level 5:                22
```
both at time `t = 1.25`.

Screenshot 2026-07-10 at 14 10 26

Screenshot 2026-07-10 at 14 10 03

### CODE REVIEW

File: examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+Initial condition to simulate a wave running towards a beach and crashing. Difficult test
+including both wetting and drying in the domain using slip wall boundary conditions.
+The bottom topography is altered to be differentiable on the domain [0,8] and
+differs from the reference below.
+
+The water height and speed functions used here, are adapted from the initial condition
+found in section 5.2 of the paper:
+  - Andreas Bollermann, Sebastian Noelle, Maria Lukáčová-Medvid’ová (2011)
+    Finite volume evolution Galerkin methods for the shallow water equations with dry beds\n
```
Comment (ranocha) [inline examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl]:
```suggestion
    Finite volume evolution Galerkin methods for the shallow water equations with dry beds
```

File: examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+###############################################################################
+# Semidiscretization of the shallow water equations
+
+equations = ShallowWaterEquations1D(gravity = 9.812)
+
+"""
+    initial_condition_beach(x, t, equations:: ShallowWaterEquations1D)
+
+Initial condition to simulate a wave running towards a beach and crashing. Difficult test
```
Comment (ranocha) [inline examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl]: Is this really how one talks about it in English? I would have expected to say that the wave breaks (instead of crashes).

File: examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+stepsize_callback = StepsizeCallback(cfl = 1.0)
+
+callbacks = CallbackSet(summary_callback, analysis_callback, alive_callback, save_solution,
+                        amr_callback, stepsize_callback)
+
+###############################################################################
+# run the simulation
+
+sol = solve(ode, SSPRK43(; stage_limiter! = positivity_limiter);
+            dt = 1.0, # solve needs some value here but it will be overwritten by the stepsize_callback
```
Comment (ranocha) [inline examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl]:
```suggestion
            dt = 1, # solve needs some value here but it will be overwritten by the stepsize_callback
```

For compatibility with `Float32` computations.

File: examples/tree_1d_dgsem/elixir_shallowwater_multilayer_beach_amr.jl [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+stepsize_callback = StepsizeCallback(cfl = 1.0)
+
+callbacks = CallbackSet(summary_callback, analysis_callback, alive_callback, save_solution,
+                        amr_callback, stepsize_callback)
+
+###############################################################################
+# run the simulation
+
+sol = solve(ode, SSPRK43(; stage_limiter! = positivity_limiter);
+            dt = 1.0, # solve needs some value here but it will be overwritten by the stepsize_callback
```
Comment (ranocha) [inline examples/tree_1d_dgsem/elixir_shallowwater_multilayer_beach_amr.jl]:
```suggestion
            dt = 1, # solve needs some value here but it will be overwritten by the stepsize_callback
```

File: src/callbacks_stage/positivity_shallow_water_dg1d.jl [hunk 5]
```diff
@@ -92,21 +237,149 @@ function limiter_shallow_water!(u, threshold::Real, variable,
...
+
+            # Iterate over the children of the current element to determine a joint limiting coefficient `theta`
+            for new_element_id in element_ids_new[idx]:(element_ids_new[idx] + 2^ndims(mesh) - 1)
+                # determine minimum value
+                value_min = typemax(eltype(u))
+                for i in eachnode(dg)
+                    u_node = get_node_vars(u, equations, dg, i, new_element_id)
+                    value_min = min(value_min, variable(u_node, equations)[m])
+                end
+                value_min < threshold - eps() || continue # Detect if limiting is necessary
```
Comment (ranocha) [inline src/callbacks_stage/positivity_shallow_water_dg1d.jl:316]: This is a hard-coded `Float64` constant `eps()` instead of something adapted to the floating-point type in use. We can continue with this version for now, but may add a TODO note and open an issue to think about other floating-point types.

File: src/callbacks_stage/positivity_shallow_water_dg1d.jl [hunk 6]
```diff
@@ -92,21 +237,149 @@ function limiter_shallow_water!(u, threshold::Real, variable,
...
+    Trixi.@threaded for element in element_ids_new
+        # Limit layerwise
+        for m in eachlayer(equations)
+            # determine minimum value
+            value_min = typemax(eltype(u))
+            for i in eachnode(dg)
+                u_node = get_node_vars(u, equations, dg, i, element)
+                value_min = min(value_min, variable(u_node, equations)[m])
+            end
+            value_min < threshold - eps() || continue # Detect if limiting is necessary
```
Comment (ranocha) [inline src/callbacks_stage/positivity_shallow_water_dg1d.jl:371]: Same here

**Review by ranocha — CHANGES_REQUESTED** [review]
Thanks a lot!

File: examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl (hunk 2 above)
Comment (andrewwinters5000) [reply examples/tree_1d_dgsem/elixir_shallowwater_beach_amr.jl]: True, this is weird English. I will change this here and in the other elixir.

File: src/callbacks_stage/positivity_shallow_water_dg1d.jl (hunk 5 above)
Comment (andrewwinters5000) [reply src/callbacks_stage/positivity_shallow_water_dg1d.jl:316]: Yes, there are several spots in the limiting that rely on `Float64`. I can add appropriate TODO statements. I think we already ave such statements in the default values of things like `threshold.limiter`. We were already aware that most of the wet/dry framework relied on double precision in its current incarnation.

File: examples/tree_1d_dgsem/elixir_shallowwater_multilayer_beach_amr.jl [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+Initial condition to simulate a wave propagating toward a beach and breaking. Difficult test
+including both wetting and drying in the domain using slip wall boundary conditions.
+The bottom topography is altered to be differentiable on the domain [0,8] and
+differs from the reference below.
+
+The water height and speed functions used here, are adapted from the initial condition
+found in section 5.2 of the paper:
+  - Andreas Bollermann, Sebastian Noelle, Maria Lukáčová-Medvidová (2011)
+    Finite volume evolution Galerkin methods for the shallow water equations with dry beds\n
```
Comment (ranocha) [inline examples/tree_1d_dgsem/elixir_shallowwater_multilayer_beach_amr.jl]:
```suggestion
    Finite volume evolution Galerkin methods for the shallow water equations with dry beds
```

**Review by ranocha — APPROVED** [review]
Thanks!
